### PR TITLE
Run the TSB in a separate pod

### PR DESCRIPTION
**Pull request by deads2k**
This completes the work necessary to run the TSB in a pod.  We should continue to remove the cruft from the old path, but I've included a template which creates a running TSB server in the `openshift-template-service-broker` namespace.

@bparees @jim-minter

### CODE REVIEW

**Comment by deads2k** [comment]
Build your images and do:

```
oc adm new-project  openshift-template-service-broker
oc process -f test/testdata/templateservicebroker/templateservicebroker-template.yaml | oc create -f -
```

@sdodson What does it take to do the equivalent in the installer?

**Comment by sdodson** [comment]
> @sdodson What does it take to do the equivalent in the installer?

Whomever owns TSB should work with @ewolinetz to contribute TSB installation automation to openshift-ansible. Is this optional? Is this installed by default when Service Catalog is enabled? Fully supported in 3.7? need to answer those sort of questions too.

**Comment by deads2k** [comment]
This won't make non-conflicting conflicts

File: pkg/openservicebroker/server/apiserver.go [hunk 1]
```diff
@@ -87,11 +90,55 @@ func (c completedTemplateServiceBrokerConfig) New(delegationTarget genericapiser
...
+			c.TemplateInformers.Template().InternalVersion().Templates(),
+			c.TemplateNamespaces,
+		)
+		// make sure no one else uses it
+		c.TemplateInformers = nil
+
+	} else {
+		// we're running in cluster
+		// in this case we actually want to construct our own template informer
+		// eventually the config value drops out as it isn't supported server side
```
Comment (bparees) [inline pkg/openservicebroker/server/apiserver.go]: "config value" refers to c.TemplateNamespaces ?

Comment (deads2k) [inline pkg/openservicebroker/server/apiserver.go]: > "config value" refers to c.TemplateNamespaces ?

tempateinformers

Comment (bparees) [inline pkg/openservicebroker/server/apiserver.go]: so where does the template informer get its config from at that point?  and is that something one of your other pulls changes?

File: pkg/template/servicebroker/servicebroker.go [hunk 2]
```diff
@@ -64,6 +64,47 @@ func DeprecatedNewBrokerInsideAPIServer(privilegedKubeClientConfig restclient.Co
...
+
+	internalKubeClient, err := kclientset.NewForConfig(saKubeClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	externalKubeClient, err := kclientsetexternal.NewForConfig(saKubeClientConfig)
+	if err != nil {
+		return nil, err
+	}
+	extrouteclientset, err := extrouteclientset.NewForConfig(saKubeClientConfig)
```
Comment (bparees) [inline pkg/template/servicebroker/servicebroker.go]: externalRouteClient (align w/ other names+case)

Comment (deads2k) [inline pkg/template/servicebroker/servicebroker.go]: > externalRouteClient (align w/ other names+case)

pre-existing.  happy to do it in a follow on.

File: pkg/openservicebroker/server/apiserver.go (hunk 1 above)
Comment (deads2k) [inline pkg/openservicebroker/server/apiserver.go]: > so where does the template informer get its config from at that point? and is that something one of your other pulls changes?

5 lines down in this pull.

Comment (bparees) [inline pkg/openservicebroker/server/apiserver.go]: i'm lost.

In the case where this code is running in a pod (not in the master), you're currently standing up an informer using a client that's created using the in cluster config.

the comment says "eventually the config value drops out".  What is meant by eventually, are you referring to a future evolution of this code(my original interpretation), or are you referring to some downstream codepath that dumps the value?  And if the latter, what is meant by "drops out"?

Comment (deads2k) [inline pkg/openservicebroker/server/apiserver.go]: > the comment says "eventually the config value drops out". What is meant by eventually, are you referring to a future evolution of this code(my original interpretation),

Future evolution removes this code.  See https://github.com/openshift/origin/pull/15684/commits/e27e5d5d252e655485950d13c367e51e147b6be5#diff-e73d057d1e253024c42b0fec37e12420L59

**Comment by deads2k** [comment]
```
Extended.[Conformance][networking][router] The HAProxy router should support reencrypt to services backed by a serving certificate automatically
Extended.[Conformance][networking][router] router headers The HAProxy router should set Forwarded headers appropriately
Extended.[Conformance][networking][router] openshift router metrics The HAProxy router should expose a health check on the metrics port
```
/retest

**Comment by deads2k** [comment]
Router again. I don't see the relationship. I'll try once more

/retest

**Comment by soltysh** [comment]
https://github.com/openshift/origin/issues/8571

/test extended_conformance_install_update

File: pkg/openservicebroker/cmd/server/start.go [hunk 3]
```diff
@@ -13,10 +13,13 @@ import (
...
 
-const defaultEtcdPathPrefix = "/registry/templateservicebroker.openshift.io"
-
 type TemplateServiceBrokerServerOptions struct {
-	RecommendedOptions *genericoptions.RecommendedOptions
+	// we don't have any storage, so we shouldn't use the recommended options
+	SecureServing  *genericoptions.SecureServingOptions
+	Authentication *genericoptions.DelegatingAuthenticationOptions
+	Authorization  *genericoptions.DelegatingAuthorizationOptions
+	Audit          *genericoptions.AuditOptions
```
Comment (soltysh) [inline pkg/openservicebroker/cmd/server/start.go]: I'm guessing this is making our audit harder, of course. Although the advanced audit is designed to work with multiple api server I'm not 100% sure if that was tested ;) Btw. do we care about upstream audit options here. We haven't decided yet which parts of the advanced audit we will enable and this is literally taking what upstream has.

**Pull request merged**